File: 3Sum.cpp
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> solution;
        sort(nums.begin(), nums.end());
        int n = nums.size();
        int l = 0;
        int r = n-1;

        for (int i = 0; i < n; i++) {
            if (i > 0 && nums[i] == nums[i-1]) { continue; }
            if (nums[i] > 0) { break; }
            int target = -nums[i];
            l = i + 1;
            r = n - 1;

            while (l < r) {
                if (nums[l] + nums[r] < target) {
                    l++;
                } else if (nums[l] + nums[r] > target) {
                    r--;
                } else if (nums[l] + nums[r] == target){
                    vector<int> asol;
                    asol.push_back(nums[i]);
                    asol.push_back(nums[l]);
                    asol.push_back(nums[r]);
                    solution.push_back(asol);
                    l++;
                    while (nums[l] == nums[l-1] && l < r) {
                        l++;
                    } 
                }
            }
        }
        return solution;
    }
};

int main() {}

File: Blind75/3sum.py
from typing import List


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        result = []

        for i, num in enumerate(nums):
            if num > 0:
                break
            if i > 0 and num == nums[i - 1]:
                continue

            target = 0 - num
            l = i + 1
            mySet = set()
            added = set()
            while l < len(nums):
                if nums[l] in mySet:
                    if (num, nums[l], target - nums[l]) not in added:
                        result.append([num, nums[l], target - nums[l]])
                        added.add((num, nums[l], target - nums[l]))
                mySet.add(target - nums[l])
                l += 1

        return result
